### ISSUE TITLE
Ability to leave the connection open

### ISSUE THREAD

**Issue by halostatue** [ruby-ldap/ruby-net-ldap#11]
Originally submitted on RubyForge by James Hewitt as #23425 on 2009-01-03:

I'd like to be able to leave an LDAP connection open.

I know the library doesn't want me to, but I would like to be able to decide what to do with my own network connections.

In some cases, leaving the ldap connection open makes writing my application a lot easier, for example, where I want to only use one ldap connection for each request to a rails app, but don't want to have to wrap the whole request in an LDAP.open. Or even to have an ldap connection pool, as you would a mysql connection pool.

I would suggest:
1. Altering LDAP.open to open the connection and leave it open if a block is not given.
2. Adding a close method to close connections.

**Comment by sonOfRa** [comment]
Sorry to be replying to such an old issue, but I stumbled across this issue while considering a rewrite of some of our LDAP management tools:

https://gist.github.com/sonOfRa/9f2474cf5cd3586d3061f7bf0669c9e7

tl;dr: In order to have accountability, LDAP actions are performed by the user who is logged in, and not by some maintenance user. Currently, we cache the password in memory so the user doesn't have to re-enter it for every LDAP action. It would be great to have connection reuse so this password storing can be eliminated.

**Comment by grawity** [comment]
@sonOfRa:

> LDAP actions are performed by the user who is logged in, and not by some maintenance user.

You could log in using the maintenance user's credentials but with the human operator's _authzid_, no?

**Comment by sonOfRa** [comment]
I have not considered this actually, thanks for the suggestion.

Based on my quick readup on how _authzid_ works, I can see these problems:

1. Now I need to store the password for such a management user on disk _persistently_ somehow
2. The management user needs to be able to "become" any user, even actual administrative users, so that user has (transitively) full write rights to the entire tree.

I'm not sure whether this is actually better than the old way, because in the old way, I only store passwords in _memory_ for the duration of the user's session, now I'd store a single password with (essentially) full write access on disk. Of course, I may have misunderstood the authzid concept completely, in which case these points probably don't apply.
